cpp/src/reshape/byte_cast.cu

```diff
@@ -0,0 +1,169 @@
+/*
+ * Copyright (c) 2020, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include <cudf/detail/reshape.hpp>
+#include <cudf/strings/detail/utilities.hpp>
+
+#include "cudf/column/column_device_view.cuh"
+#include "cudf/detail/nvtx/ranges.hpp"
+#include "cudf/replace.hpp"
+#include "cudf/strings/detail/utilities.cuh"
+#include "cudf/utilities/traits.hpp"
+#include "cudf/utilities/type_dispatcher.hpp"
+
+namespace cudf {
+namespace {
+struct byte_list_conversion {
+  /**
+   * @brief Function object for converting primitive types and string columns to lists of bytes,
+   * mimics Spark's cast to binary type.
+   */
+  struct FlipEndianness {
+    char* d_chars;
+    const char* d_data;
+    uint32_t mask;
+    __device__ void operator()(int byte_index)
+    {
+      d_chars[byte_index] = d_data[byte_index + mask - ((byte_index & mask) << 1)];
+    }
+  };
+
+  template <typename T,
+            std::enable_if_t<!std::is_integral<T>::value and !is_floating_point<T>()>* = nullptr>
+  std::unique_ptr<column> operator()(column_view const& input_column,
+                                     flip_endianness configuration,
+                                     rmm::mr::device_memory_resource* mr,
+                                     cudaStream_t stream) const
+  {
+    CUDF_FAIL("Unsupported non-numeric and non-string column");
+  }
+
+  template <typename T, std::enable_if_t<is_floating_point<T>()>* = nullptr>
+  std::unique_ptr<column> operator()(column_view const& input_column,
+                                     flip_endianness configuration,
+                                     rmm::mr::device_memory_resource* mr,
+                                     cudaStream_t stream) const
+  {
+    size_type num_output_elements = input_column.size() * cudf::size_of(input_column.type());
+
+    auto begin          = thrust::make_constant_iterator(cudf::size_of(input_column.type()));
+    auto offsets_column = cudf::strings::detail::make_offsets_child_column(
+      begin, begin + input_column.size(), mr, stream);
+
+    auto byte_column = make_numeric_column(
+      data_type{type_id::UINT8}, num_output_elements, mask_state::UNALLOCATED, stream, mr);
+    auto d_chars = byte_column->mutable_view().data<char>();
+
+    rmm::device_buffer null_mask = copy_bitmask(input_column, stream, mr);
+    auto normalized              = normalize_nans_and_zeros(input_column);
```

Comment:
Much of the code is same for integral and floating point columns except this one extra step. Recommend merging most of the code. Perhaps a separate function in this anonymous namespace that indicates that it works for all fixed width types. Also, don't you want timestamps?

Reply:
If you see a use case and clear way to translate the timestamp data into a byte list I can add the functionality. Timestamp translation to another data format is tricky because you end up dropping the unit of the timestamp. I think a user should be drop the time unit themselves rather than integrate the loss of that information into other functions.

Reply:
Sure, maybe not timestamps, but these two can still be consolidated.

Reply:
Code has been changed to the point where it's useful for them to be split again, the 2 types no longer use the same kernels.

Reply:
Has now been re-combined